packages/controllers/src/snaps/SnapController.ts

```diff
@@ -1530,12 +1534,39 @@ export class SnapController extends BaseController<
       }
 
       if (this.state.snaps[snapId].status === SnapStatus.installing) {
-        throw new Error(`Snap "${snapId}" has not been started yet.`);
+        throw new Error(
+          `Snap "${snapId}" is currently being installed. Please try again later.`,
+        );
       }
 
-      if (!handler && this.isRunning(snapId) === false) {
-        // cold start
-        await this.startSnap(snapId);
+      if (this.isRunning(snapId) === false) {
+        if (handler) {
+          throw new Error(
+            'This snap should not have a handler in its current state. This is a bug, please report it.',
+          );
+        }
+
+        if (!startPromise) {
+          startPromise = this.startSnap(snapId);
```

Comment:
Do we not need a potential `startPromise` for each snap id? With this, we'll have a problem if multiple snaps are being requested at once, right?

Reply:
A call to `getRpcMessageHandler` produces an `rpcHandler` that is specific to a snap, which means each snap would have its own `startPromise`, so I think we're fine here.

Reply:
@FrederikBolding Apologies! On second thought, you were actually right about needing a `startPromise` for every snap id, pushed up a commit that sets up a map. Test coming in a few as well.

Reply:
Hmm. How come we don't need a request queue for each snap too?

Reply:
The request queue has an internal map for each snap, unless you're saying that won't suffice?

Reply:
Oh yes, you are right, my bad! Should we combine the `startPromises` and the RequestQueue somehow? 🤔

Reply:
I think it is fine to leave them separate, seems like it would be confusing to have the two together.

Reply:
@hmalik88 Perhaps we could move the promises to `_snapsRuntimeData` then? Just so we don't have too many Maps scattered 😄 

`RequestQueue` could be on that as well if it was only representative of a single snap, but up to you!

Reply:
Hmm I like that idea, I think I had mentioned something similar before. Thoughts? @rekmarks

Reply:
Consolidating shared data is good, but both the request queue and the start promise can be regarded as internal implementation details of `getRpcMessageHandler`. If we ever need to access either the promise or the request queue elsewhere we can move it to the runtime data.

Reply:
I'm fine with that too 👍